### PR TITLE
fqdn: Fix duplicate CIDRSet generation

### PR DESCRIPTION
**Pull request by tgraf**
When using match patterns, the FQDN cache returns a map of results, each
containing a list of IPs. This can lead to duplicate IPs inserted into the list
of rules.

Fixes: #6541

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6542)

### CODE REVIEW

File: pkg/fqdn/helpers.go [hunk 1]
```diff
@@ -52,6 +53,9 @@ func injectToCIDRSetRules(rule *api.Rule, cache *DNSCache, reMap *regexpmap.Rege
 	for egressIdx := range rule.Egress {
 		egressRule := &rule.Egress[egressIdx]
 
+		// Build an IP collection to remove all duplicates
+		allIPs := ip.NewCollection()
```
Comment (raybejjani) [inline pkg/fqdn/helpers.go]: Could you make this a `[]net.IP` and use `keepUniqueIPs` (https://github.com/cilium/cilium/blob/master/pkg/ip/ip.go#L730) on line 111, where we finally add toCIDRSet rules?

**Review by raybejjani — CHANGES_REQUESTED** [review]
I'd prefer to use `keepUniqueIPs` (https://github.com/cilium/cilium/blob/master/pkg/ip/ip.go#L730) unless there was another reason you wanted `ip.Collection`. This might have to be backported so it would be a little simpler then.

Edit: I also noticed the unit tests failed. They are brittle, but assume the output list is sorted, which `keepUniqueIPs` keeps.

**Comment by tgraf** [comment]
test-me-please

**Review by raybejjani — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

File: pkg/fqdn/helpers.go [hunk 2]
```diff
@@ -99,10 +103,15 @@ func injectToCIDRSetRules(rule *api.Rule, cache *DNSCache, reMap *regexpmap.Rege
...
 					emitted[name] = append(emitted[name], ips...)
-					egressRule.ToCIDRSet = append(egressRule.ToCIDRSet, api.IPsToCIDRRules(ips)...)
+					allIPs = append(allIPs, ips...)
 				}
 			}
 		}
+
+		// Only overwrite ToCIDRSet for rules with FQDN elements
+		if len(allIPs) > 0 {
+			egressRule.ToCIDRSet = api.IPsToCIDRRules(ip.KeepUniqueIPs(allIPs))
```
Comment (raybejjani) [inline pkg/fqdn/helpers.go:112]: oh, one last thing. The comment on`injectToCIDRSetRules` already existing rules are preserved. We should remove that line (44) above.

**Review by aanm — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**